File: 백준/Silver/13565. 침투/침투.py
import sys
sys.setrecursionlimit(1000010)

def dfs(r, c, l, m, n, vis):
    if r == m - 1:
        return "YES"
    
    vis[r][c] = 1

    dr = [1, 0, -1, 0]
    dc = [0, 1, 0, -1]

    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        if (0 <= nr < m and 0 <= nc < n and 
            l[nr][nc] == '0' and not vis[nr][nc]):
            if dfs(nr, nc, l, m, n, vis) == "YES":
                return "YES"
    
    return "NO"

def main():
    m, n = map(int, input().split())
    l = [input() for _ in range(m)]
    vis = [[0] * n for _ in range(m)]
    for j in range(n):
        if l[0][j] == '0' and not vis[0][j]:
            if dfs(0, j, l, m, n, vis) == "YES":
                print("YES")
                exit(0)
    print("NO")
    
if __name__ == '__main__':
    main()

File: 백준/Silver/13565. 침투/침투.kt
package org.example

val dx = intArrayOf(-1, 1, 0, 0)
val dy = intArrayOf(0, 0, -1, 1)

fun dfs(graph: Array<IntArray>, x: Int, y: Int): Boolean {
    if (x == graph.size - 1) {
        return true
    }
    graph[x][y] = 1

    for (i in 0..3) {
        val nx = x + dx[i]
        val ny = y + dy[i]
        if (nx in graph.indices && ny in graph[0].indices && graph[nx][ny] == 0) {
            if (dfs(graph, nx, ny)) {
                return true
            }
        }
    }
    return false
}

fun main() {
    val br = System.`in`.bufferedReader()

    val (m, n) = br.readLine().split(" ").map { it.toInt() }
    val graph = Array(m) { br.readLine().map { it.digitToInt() }.toIntArray() }

    for (j in 0 until n) {
        if (graph[0][j] == 0) {
            if (dfs(graph, 0, j)) {
                println("YES")
                return
            }
        }
    }
    println("NO")
    br.close()
}